### ISSUE TITLE
propagate-bv-bounds tactic consuming large amounts of memory, possibly stuck in a loop

### ISSUE THREAD

**Issue by delcypher** [Z3Prover/z3#1078]
I'm using the `propagate-bv-bounds` tactic as a pre-processing step and I've encountered a benchmark that causes Z3 to consumes huge amounts (> 20GiB) of memory and CPU time when I try to apply this tactic.

The benchmark is [apply-intSqRoot.smt2.zip](https://github.com/Z3Prover/z3/files/1075780/apply-intSqRoot.smt2.zip). This is taken from the SMTLIB QF_BV benchmark suite.

The benchmark is `QF_BV/bmc-bv/intSqRoot.smt2` from the  SMTLIB QF_BV benchmark suite but I've replaced `(check-sat)` with `(apply (and-then simplify propagate-bv-bounds))`.

Is this expected behaviour?

**Comment by NikolajBjorner** [comment]
It looks like it is the following:    
         expr_set* get_expr_vars(expr* t) 
creates intermediary sets. The sets are used for checking containment in the function
        virtual bool may_simplify(expr* t)
in bv_bounds_tactic.
It then applies context simplification, which can also be super-expensive in its current form (hold on in suspense for an improvement coming your way after the all important POPL deadline).

@nunoplopes, can you take a look at my changes to address memory overhead?

**Comment by delcypher** [comment]
@NikolajBjorner Thanks for the quick fix. Memory usage seems to have gone down but CPU usage is still very high and is taking a very long time to run on this benchmark. I gave it ~5 minutes to run and it didn't finish so I killed Z3.

**Comment by nunoplopes** [comment]
I wouldn't use this tactic right now. The algorithm is too expensive for big formulas.
We have something else in the pipeline to replace this (fingers crossed).
@delcypher I'm curious, why are you running this tactic? Did it improve any of your benchmarks?

**Comment by rainoftime** [comment]
@delcypher  Agree with nunoplopes. `propagate-bv-bounds` tactic is too expensive

**Comment by delcypher** [comment]
@nunoplopes I'm currently building an experimental solver on top of Z3 that uses an incomplete method that can only prove sat. As such I wanted to use as many cheap pre-processing steps as possible so that I can determine a query is unsat in fairly trivial cases.

There isn't any detailed documentation on what each tactic does or what their complexity is so I resorted to looking at the list of available tactics and trying out tactics that sounded helpful based on their name. The `propagate-bv-bounds` tactic sounded like it would simplify contradictory bounds to false. My initial experimentation showed that it could do this but these were on small examples. I only recently started trying on much larger examples and ran into performance problems.

For now I've stopped using this tactic. If I get time I'll implement my own approach that just looks for very simple contradictory bounds which is all I wanted to do.
